src/engagement_db_to_analysis/rapid_pro_advert_functions.py

```diff
@@ -0,0 +1,271 @@
+
+from core_data_modules.cleaners import Codes
+from core_data_modules.logging import Logger
+from core_data_modules.analysis import analysis_utils, AnalysisConfiguration as core_data_analysis_config
+
+from src.engagement_db_to_analysis.membership_group import (get_membership_groups_data)
+from src.engagement_db_to_analysis.cache import AnalysisCache
+
+from src.pipeline_configuration_spec import *
+
+
+log = Logger(__name__)
+
+CONSENT_WITHDRAWN_KEY = "consent_withdrawn"
+
+
+def _generate_weekly_advert_and_opt_out_uuids(participants_by_column, analysis_config,
+                                     google_cloud_credentials_file_path, membership_group_dir_path):
+    '''
+    Generates sets of weekly advert and  opt_out UUIDs to advertise to in rapid_pro.
+
+    :param participants_by_column: Participants column view TracedData object to generate the uuids from.
+    :type participants_by_column: core_data_modules.traced_data.TracedData
+    :param analysis_config: Configuration for the export.
+    :type analysis_config: src.engagement_db_to_analysis.configuration.AnalysisConfiguration
+    :param google_cloud_credentials_file_path: Path to the Google Cloud service account credentials file to use to
+                                               access the credentials bucket.
+    :param membership_group_dir_path: Path to directory containing de-identified membership groups CSVs containing membership groups data
+                        stored as `avf-participant-uuid` column.
+    :type: membership_group_dir_path: str
+    :return opt_out_uuids and weekly_advert_uuids : Set of opted out and weekly advert uuids.
+    :rtype opt_out_uuids & weekly_advert_uuids: set of str
```

Comment:
```suggestion
    :return: Tuple of (set of opted out uuids, set of weekly advert uuids)
    :rtype: tuple of (set of str, set of str)
```
(Not sure if you need the tuple or if you can just say (set of str, set of str) to make editors understand the types).